code/game/machinery/atm_ret_field.dm

```diff
@@ -0,0 +1,246 @@
+	alwaysactive = TRUE
+	active_power_usage = 2000
+
+/obj/machinery/atmospheric_field_generator/perma/impassable
+	desc = "An older model of ARF-G that generates an impassable retention field whilst powered. This model's field is quite power-efficient, but will still drop from loss of power or electromagnetic interference.<br><br>Note: prolonged immersion in active atmospheric retention fields may have negative long-term health consequences."
+	active_power_usage = 1500
+	field_type = /obj/structure/atmospheric_retention_field/impassable
+
+/obj/machinery/atmospheric_field_generator/attackby(obj/item/W as obj, mob/user as mob)
+	if(W.is_crowbar() && isactive)
+		if(!src) return
+		to_chat(user, "<span class='warning'>You can't open the ARF-G whilst it's running!</span>")
+		return
+	if(W.is_crowbar() && !isactive)
+		if(!src) return
+		to_chat(user, "<span class='notice'>You [hatch_open? "close" : "open"] \the [src]'s access hatch.</span>")
+		hatch_open = !hatch_open
+		update_icon()
+		if(alwaysactive && wires_intact)
+			generate_field()
+		return
+	if(hatch_open && W.is_multitool())
+		if(!src) return
+		to_chat(user, "<span class='notice'>You toggle \the [src]'s activation behavior to [alwaysactive? "emergency" : "always-on"].</span>")
+		alwaysactive = !alwaysactive
+		update_icon()
+		return
+	if(hatch_open && W.is_wirecutter())
+		if(!src) return
+		to_chat(user, "<span class='warning'>You [wires_intact? "cut" : "mend"] \the [src]'s wires!</span>")
+		wires_intact = !wires_intact
+		update_icon()
+		return
+	if(hatch_open && istype(W,/obj/item/weldingtool))
+		if(!src) return
+		var/obj/item/weldingtool/WT = W
+		if(!WT.isOn())
+			return
+		if(!WT.remove_fuel(0,user))
+			to_chat(user, "<span class='warning'>You need more fuel to complete this task.</span>")
+			return
+		user.visible_message("[user] starts to disassemble \the [src].", "You start to disassemble \the [src].")
+		playsound(src, WT.usesound, 50, 1)
+		if(do_after(user,15 * W.toolspeed))
+			if(!src || !user || !WT.remove_fuel(5, user)) return
+			to_chat(user, "<span class='notice'>You fully disassemble \the [src]. There were no salvageable parts.</span>")
+			qdel(src)
+		return
+
+/obj/machinery/atmospheric_field_generator/perma/Initialize()
+	. = ..()
+	generate_field()
+
+/obj/machinery/atmospheric_field_generator/update_icon()
+	if(stat & BROKEN)
+		icon_state = "arfg_broken"
+	else if(hatch_open && wires_intact)
+		icon_state = "arfg_open_wires"
+	else if(hatch_open && !wires_intact)
+		icon_state = "arfg_open_wirescut"
+	else if(isactive)
+		icon_state = "arfg_on"
+	else
+		icon_state = "arfg_off"
+
+/obj/machinery/atmospheric_field_generator/power_change()
+	var/oldstat
+	. = ..()
+	if(!(stat & NOPOWER))
+		ispowered = 1
+		update_icon()
+		if(alwaysactive || wasactive)	//reboot our field if we were on or are supposed to be always-on
+			generate_field()
+	if(stat != oldstat && isactive && (stat & NOPOWER))
+		ispowered = 0
+		disable_field()
+		update_icon()
```

Comment:
Are you calling `update_icon()` every time you `generate_field()` or `disable_field()`? Should those procs be the thing that calls `update_icon()`?